### PR TITLE
Implement Custom Reporters

### PR DESCRIPTION
**Pull request by TheBrokenRail**
### Proposed changes

_Adds Code for custom reporters_

### Notes

_I did not add another definition block because they do the same thing, if there just isn't a return value it just returns 'undefined'_

### CODE REVIEW

File: src/blocks/scratch3_procedures.js [hunk 1]
```diff
@@ -14,25 +15,49 @@ Scratch3ProcedureBlocks.prototype.getPrimitives = function () {
...
+        util.startProcedure(procedureCode, util.stackFrame.thread);
+    }
+};
+
+Scratch3ProcedureBlocks.prototype.callReturn = function (args, util) {
+    if (!util.stackFrame.executed) {
+        var procedureCode = args.mutation.proccode;
+        var paramNames = util.getProcedureParamNames(procedureCode);
+        util.stackFrame.thread = this.thread;
+        this.thread = this.thread + 1;
```
Comment (dekrain) [inline src/blocks/scratch3_procedures.js]:
``` JavaScript
this.thread++;
```

File: src/blocks/scratch3_procedures.js [hunk 2]
```diff
@@ -14,25 +15,49 @@ Scratch3ProcedureBlocks.prototype.getPrimitives = function () {
...
+    util.stackFrame.done = true;
 };
 
 Scratch3ProcedureBlocks.prototype.callNoReturn = function (args, util) {
     if (!util.stackFrame.executed) {
         var procedureCode = args.mutation.proccode;
         var paramNames = util.getProcedureParamNames(procedureCode);
+        util.stackFrame.thread = this.thread;
+        this.thread = this.thread + 1;
         for (var i = 0; i < paramNames.length; i++) {
```
Comment (dekrain) [inline src/blocks/scratch3_procedures.js:35]: ...and here.

File: src/engine/execute.js [hunk 3]
```diff
@@ -198,6 +227,9 @@ var execute = function (sequencer, thread) {
         getParam: function (paramName) {
             return thread.getParam(paramName);
         },
+        setProcedureReturn: function (value) {
+            thread.stackFrames[0].executionContext.return = value;
```
Comment (dekrain) [inline src/engine/execute.js:231]:
``` JavaScript
currentStackFrame.executionContext.return = value;
```

**Review by dekrain — CHANGES_REQUESTED** [review]

**Comment by dekrain** [comment]
`src/engine/execute.js#231`:

``` javascript
currentStackFrame.executionContext.return = value;
```

File: src/engine/execute.js (hunk 3 above)
Comment (TheBrokenRail) [reply src/engine/execute.js:231]: Wouldn't that cause the value to be set to the report block instead of the top block (hat)

Comment (dekrain) [reply src/engine/execute.js:231]: So why you added

``` JavaScript
for (; k < this.runtime.threads[k].stackFrames.length; k++)
```

?

Comment (TheBrokenRail) [reply src/engine/execute.js:231]: That runs from the calling block and not the hat block so it has to "find" the stackFrame

Comment (dekrain) [reply src/engine/execute.js:231]: Shouldn't it be

``` JavaScript
var stack = this.runtime.threads[k].stackFrames[0];
```

?

**Comment by CosmicWebServices** [comment]
@TheBrokenRail you should finish this

**Comment by TheBrokenRail** [comment]
I can't test this until the scratch blocks PR is merged

**Comment by towerofnix** [comment]
You can probably figure out how to get a local copies of scratch-blocks and scratch-vm to work together, but I don't know if there's any official documentation on that.

**Comment by TheBrokenRail** [comment]
I will close this until a design for implementation is more finalized and testable.

**Pull request closed without merging**